File: meme_inator_back/apps/profiles/application/usecases/get_profile_posts_usecase.py
from typing import Dict, Any, Optional
from uuid import UUID
from apps.profiles.domain.usecases.iget_profile_posts_usecase import IGetProfilePostsUsecase
from apps.profiles.domain.irepositories.i_profile_repository import IProfileRepository
from core.results import Result

class GetProfilePostsUsecase(IGetProfilePostsUsecase):
    def __init__(self, profile_repo: IProfileRepository):
        self.profile_repo = profile_repo

    def execute(self, username: str, cursor: Optional[str] = None, page_size: int = 10) -> Result[Dict[str, Any]]:
        # The repository returns a Result containing a dict with 'posts' and 'next_cursor'
        return self.profile_repo.get_profile_posts(username, cursor, page_size)
    

File: meme_inator_front/lib/features/profiles/domain/usecases/get_profile_posts_usecase.dart
// lib/features/profiles/domain/usecases/get_profile_posts_usecase.dart
import 'package:meme_inator_front/core/results.dart';
import 'package:meme_inator_front/features/post/posts_mapper.dart';
import 'package:meme_inator_front/features/profiles/domain/entities/vos/get_profile_posts_request_vo.dart';
import 'package:meme_inator_front/features/profiles/domain/entities/vos/profile_posts_response_vo.dart';
import 'package:meme_inator_front/features/profiles/domain/interfaces/iprofile_repository.dart';

/// Simple use case for fetching profile posts
/// Returns ProfilePostsResponseVo containing posts and next cursor
class GetProfilePostsUsecase {
  final IProfileRepository repository;

  GetProfilePostsUsecase({required this.repository});

  Future<Result<ProfilePostsResponseVo>> execute({
    required GetProfilePostsRequestVo request,
  }) async {
    try {
      final result = await repository.getProfilePosts(
        username: request.username,
        cursor: request.cursor,
        pageSize: request.pageSize,
      );

      return result.match(
        ok: (data) {
          // Convert the raw map response to our value object
          final posts = (data['posts'] as List)
              .map((post) => PostsMapper.fromJson(post as Map<String, dynamic>))
              .toList();
          final nextCursor = data['next_cursor'] as String?;
          
          return Ok(ProfilePostsResponseVo(
            posts: posts,
            nextCursor: nextCursor,
          ));
        },
        notOk: (notOk) => NotOk(
          statusCode: notOk.statusCode,
          message: notOk.message,
          staticMessage: notOk.staticMessage,
        ),
        error: (error) => Error(
          statusCode: error.statusCode,
          message: error.message,
          staticMessage: error.staticMessage,
          exception: error.exception,
        ),
      );
    } catch (e) {
      return Error.fromException(
        e is Exception ? e : Exception(e.toString()),
        message: 'Failed to get profile posts',
      );
    }
  }
}